### DISCUSSION TITLE
Questions around some inconsistencies

### DISCUSSION THREAD

**Discussion by sidekick-eimantas** [Homebrew/discussions#3765] [Everyday usage] [answered]
Hi

I just finished integrating brew into our company cli tooling, and in the process, ran into a couple of, what I perceive to be, inconsistencies that complicated the integration. I wanted to raise some questions around them so I can document them correctly on my end and maybe begin a discussion for having some of them addressed.

1. JSON schema for casks and formulas are different in the output of `brew info package --json=v2`. Why is that?
2. In `brew list` you can filter for casks by using the `--cask` flag, which matches the name of the `homebrew-cask` tap, however, `--formula` must be used for packages installed through the `core` tap. The naming disparity between tap names and brews internal naming is also reflected in the JSON schemas of `brew info` and `brew outdated`. Am I using the wrong mental model, and tap names have nothing to with formulas or casks?
3. `brew info pycharm-ce --json=v1` returns `Error: No available formula with the name "pycharm-ce".`, while it works with `--json=v2` flag. Is `--json=v1` not just a formatting flag but is also used in lookups and is trying to look up a formula?
4. Various brew command outputs like `brew list --full-name -1` prefix packages sources from taps other than `core` and `cask` with the name of the tap, but does not do the same for packages sources from `core` and `cask` taps. Is brew going to maintain `core` and `cask` as first class taps long term, or is it attempting to move away from it?

Thank you !

**Comment by Bo98** [comment, marked as answer]
> 1. JSON schema for casks and formulas are different in the output of `brew info package --json=v2`. Why is that?

The way formulae and casks work are totally different - from the way to they are defined right down to the way they are installed.

You can see this comparing how a formula is defined in code (example: https://github.com/Homebrew/homebrew-core/blob/master/Formula/wget.rb) to how a cask is (example: https://github.com/Homebrew/homebrew-cask/blob/master/Casks/pycharm-ce.rb).

There is some overlap which is reflected in the schema, e.g. `version`, `tap` etc. But there's a lot that simply don't. For example for formulae you have stable and HEAD source archive URLs as well as bottle URLs for each OS, but for casks there is just one URL used for all installations.

We've managed to hide a lot of the difference in the CLI, but open the lid even slightly and the differences become apparent.

> 2. In `brew list` you can filter for casks by using the `--cask` flag, which matches the name of the `homebrew-cask` tap, however, `--formula` must be used for packages installed through the `core` tap. The naming disparity between tap names and brews internal naming is also reflected in the JSON schemas of `brew info` and `brew outdated`. Am I using the wrong mental model, and tap names have nothing to with formulas or casks?

You can view taps as repositories. Homebrew/core is https://github.com/Homebrew/homebrew-core and Homebrew/cask is https://github.com/Homebrew/homebrew-cask. These could have been one repository if we wanted to - it's just simpler this way for us to maintain.

There can equally be other taps users may have. For example we have extra cask taps like Homebrew/cask-versions: https://github.com/Homebrew/homebrew-cask-versions. You can find taps from third-parties like `heroku/brew`: https://github.com/Heroku/homebrew-brew. Taps can contain both casks and formulae if they want to.

> 3. `brew info pycharm-ce --json=v1` returns `Error: No available formula with the name "pycharm-ce".`, while it works with `--json=v2` flag. Is `--json=v1` not just a formatting flag but is also used in lookups and is trying to look up a formula?

`--json=v1` is not support for casks. It only supports formulae. `pycharm-ce` is a cask.

In fact, cask support is the main reason for v2 existing.

> 4. Various brew command outputs like `brew list --full-name -1` prefix packages sources from taps other than `core` and `cask` with the name of the tap, but does not do the same for packages sources from `core` and `cask` taps. Is brew going to maintain `core` and `cask` as first class taps long term, or is it attempting to move away from it?

There is no plans to change anything regarding this. `Homebrew/core` and `Homebrew/cask` will continue to be the primary destinations for formulae and casks respectively.

You can still prefix with `Homebrew/(core|cask)` if you like. `brew install hello` and `brew install Homebrew/core/hello` will do the same thing.

If a formula doesn't exist in `Homebrew/core` or `Homebrew/cask`, then it will still search third party taps. So for example, if you have `heroku/brew` tapped and ran `brew install heroku` then it will still act like `brew install heroku/brew/heroku` if there's no `heroku` in `Homebrew/core` and `Homebrew/cask` (which there currently isn't). The reason we just generally promote full names for third-party taps is we cannot guarantee they contain unique names so if you have two taps with a formula of the same name then it's undefined which will be picked. `Homebrew/core` and `Homebrew/cask` packages are however guaranteed to be checked first so there isn't a concern for those (well technically the same name could exist for a formula and a cask but that's why `--formula` and `--cask` exist since this problem can exist within a single tap too).

**Reply by sidekick-eimantas** [reply]
This is all really good insight, thank you
